BJS2-78108-analytic-service-consumer

MightySonnya opened the pull request:
Добавил consumer , написал логику , проверил отправку-получения событий в кафку

MightySonnya committed:
BJS2-78108-analytic-service-consumer
Добавил consumer , написал логику , проверил отправку-получения событий в кафку

tanufa25 commented on src/main/java/faang/school/analytics/controller/AnalyticsEventController.java:
```diff
@@ -0,0 +1,26 @@
+    @PostMapping("/comment-event")
+    public void handleCommentEvent(@RequestBody AnalyticsEvent event) {
+        AnalyticsDto analyticsEvent = AnalyticsEventMapper.toDto(event);
+        analyticsEventService.processCommentEvent(event);
```
А почему в контроллере приходит сама сущность, а не dto? 
А потом эта сущность маппером в dto превращается, а сервис вызывается с пришедшей сущностью. 
Зачем тогда маппинг?
Вообще по-хорошему на вход у нас должна быть dto, потом её маппером в сущность нужно превратить, и вызвать сервис с сущностью на вход.

tanufa25 replied:
А зачем вообще это в контроллере? Это же в лисенере должно быть, и есть в лисенере.
Этот метод вообще не нужен, получается. 
В контроллере можно какую-нибудь ручку для получения аналитики, а это точно не надо.

tanufa25 commented on src/main/java/faang/school/analytics/service/AnalyticsEventService.java:
```diff
@@ -0,0 +1,30 @@
+                .receivedAt(analyticsEvent.getReceivedAt())
+                .build();
+
+        AnalyticsEvent event = AnalyticsEventMapper.toEntity(dto);
```
А зачем в сервисе собирать в dto, мапить обратно в сущность и потом уже сохранять в бд?

tanufa25 commented on src/main/java/faang/school/analytics/publisher/AbstractEventPublisher.java:
```diff
@@ -0,0 +1,29 @@
+
+@RequiredArgsConstructor
+@Slf4j
+public abstract class AbstractEventPublisher<T extends AnalyticsEvent> {
```
А у тебя паблишер как-то используется? Это же микросервис-консьюмер, тут вроде только листенер должен быть, по крайней мере, в твоей задаче точно только как слушатель

tanufa25 commented on src/main/java/faang/school/analytics/listner/CommentEventListener.java:
```diff
@@ -0,0 +1,23 @@
+            groupId = "${spring.kafka.consumer.group-id}",
+            containerFactory = "kafkaAnalyticsEventListener"
+    )
+    public void listenerEventStart(AnalyticsEvent event) {
```
AnalyticsEvent - это сущность в БД, а тут на вход ты должен получать dto, по условию задачи CommentEvent, как я понимаю. Его мапить в сущность AnalyticsEvent, добавлять в его поле EventType.POST_COMMENT, если это не было отправлено, и потом передавать эту сущность в analyticsEventService, где уже просто сохранять и не делать никакой мапингов внутри него.

tanufa25 commented on src/main/java/faang/school/analytics/config/kafka/KafkaConsumerConfig.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    private <T extends AnalyticsEvent> ConcurrentKafkaListenerContainerFactory<String, T> concurrentKafkaListenerJsonFactory() {
+        Map<String, Object> jsonFactoryConfig = new HashMap<>();
+        jsonFactoryConfig.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaProperties.getBootstrapServers());
+        jsonFactoryConfig.put(ConsumerConfig.GROUP_ID_CONFIG, kafkaProperties.getConsumer().getGroupId());
+        jsonFactoryConfig.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        jsonFactoryConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonDeserializer.class);
+        jsonFactoryConfig.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
```
Можно эти настройки сделать в application.properties

tanufa25 reviewed (commented):
Как-то всё запутано, видимо, ты не до конца разобрался с логикой происходящего.

MightySonnya committed:
BJS2-78108-analytic-service-consumer
Добавил consumer , написал логику , проверил отправку-получения событий в кафку, написал тесты

FROLOV-KI commented on src/main/java/faang/school/analytics/controller/AnalyticsEventController.java:
```diff
@@ -0,0 +1,40 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/analytics")
+public class AnalyticsEventController {
```
А зачем нам контроллер? Мы же вроде аналитику по kafka получать должны?

FROLOV-KI commented on src/main/java/faang/school/analytics/service/AnalyticsEventService.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public List<AnalyticsEvent> getAllEvents() {
+        return (List<AnalyticsEvent>) repository.findAll();
```
прям вот вообще все, а если оч много? Мб сделать возврат по пользователю, по посту и тд?

FROLOV-KI commented on src/main/resources/application.yaml:
```diff
@@ -1,4 +1,25 @@
+    topics:
+      goal-completed-topic:
+        name: goal-completed
+      subscription:
```
А это тут зачем?

FROLOV-KI commented on src/main/resources/application.yaml:
```diff
@@ -1,4 +1,25 @@
+    consumer:
+      group-id: analytic-service-group
+    topics:
+      goal-completed-topic:
```
И это?

FROLOV-KI commented on src/test/java/faang/school/analytics/controller/AnalyticsControllerTest.java:
```diff
@@ -0,0 +1,148 @@
+
+        String requestBody = objectMapper.writeValueAsString(dto);
+
+        MvcResult result = mockMvc.perform(post("/api/analytics/comment-event")
```
Мы же получаем ивент по kafkа, если уже и тестировать, то нужно тогда делать тестовый продьюсер и через него тестовые сущности закидывать в кафку и из нее получать ивенты

FROLOV-KI commented on src/main/java/faang/school/analytics/listner/CommentEventListener.java:
```diff
@@ -0,0 +1,33 @@
+    public void listenerEventStart(AnalyticsDto dto) {
+        AnalyticsEvent event = mapper.toEntity(dto);
+
+        if (event.getEventType() == null) {
```
а почему у event может не быть типа?
